### PR TITLE
[REF] base: Remove deprecated exceptions and osv

### PR DESCRIPTION
**Pull request by Julien00859**
TL;DR: remember `osv` and `except_orm` ? You can forget about them.

* Deprecated `except_orm` dropped.
* `UserError` elevated as super type of all user-related
  errors.
* Unused `DeferredException` dropped.
* Unused `QWebException` dropped (real one is in `qweb.py`).
* `MailDeliveryException` made a python exception.
* `name` legacy exception attribute made an alias of the python standard
  `args[0]` attribute and deprecated.
* `value` legacy exception attribute dropped.
* `exception_type` RPC error response key dropped.
* Deprecated `osv` module dropped.
* `--osv-memory-age-limit` cli option made an alias of
  `--transient-age-limit` and deprecated.

The `odoo.exceptions.Warning` have long been a deprecated alias to
`UserError`. It is going to be removed in a future version but first we
explicitly deprecate it with a warning.

The `odoo.exceptions.DeferredException` was a very old internal
exception, it has been removed without deprecation notice as it is never
raised.

The `odoo.exceptions.except_orm` has been a deprecated exception type
with deprecation warning for 5 years, it has been removed in favor of
UserError which becomes the super class of all user-related errors.

The `odoo.base.models.ir_mail_server.MailDeliveryException` was
inheriting `except_orm`. As it is not related to a user error but is
more of a problem an admin should take care of, the exception has
been made a Python error.

The `exception_type` JSON key in RPC error responses was holding an
hardcoded value derived from the exception type. Its usage has been
dropped in favor of the `name` JSON key that holds the precise exception
name. Again as it was hardly used in the source code (beside the crash
manager) it has been dropped without deprecation warning.

Since we are here trying to clean odoo custom exceptions, we are also
deprecating the `name` exception attribute in favor of the more standard
`args[0]` attribute.

The `name` (along with `value`) were two attributes used to raise
`except_orm` exceptions before the introduction of `UserError`,
`AccessError` and related exceptions. The `name` attribute, at the time,
was holding the exception type/title. Nowadays it contains the error
message. The `value` attribute, at the time, was holding the error
message. Nowadays it is no more used.

The `osv` module contains very old deprecated aliases. There is no
simple way to log a deprecation warning for osv, osv_memory and
osv_abstract but as they have not been in use for ages, they have been
removed too. To be consistent, the `--osv-memory-age-limit` cli option
has been made a deprecated alias to the `--transient-age-limit`.

Task: 2187728

### CODE REVIEW

**Comment by Julien00859** [comment]
So next time (say in v15) we want to remove it, we can :)

File: odoo/exceptions.py [hunk 1]
```diff
@@ -47,7 +47,13 @@ def __init__(self, msg):
 
 
 # deprecated due to collision with builtins, kept for compatibility
-Warning = UserError
+class Warning(UserError):
+    def __init__(self, *args, **kwargs):
+        _logger.warning(
+            "Warning is a deprecated alias to UserError and is going "
+            "to be removed in a future update.",
+            exc_info=DeprecationWarning(), stack_info=True)
```
Comment (xmo-odoo) [inline odoo/exceptions.py]: Why not `warnings.warn(..., DeprecationWarning)`.

Comment (Julien00859) [reply odoo/exceptions.py]: Because warnings are opt-in nobody enable

Comment (xmo-odoo) [reply odoo/exceptions.py]: > Because warnings are opt-in nobody enable

https://github.com/odoo/odoo/blob/60b0f6da3420e0bb0607bd73abb4d0a1f5a0a972/odoo/netsvc.py#L131-L132

Comment (Julien00859) [reply odoo/exceptions.py]: Oh nice !

**Comment by Julien00859** [comment]
@xmo-odoo don't forget me

**Comment by xmo-odoo** [comment]
> python standard message attribute

`message` has not been a standard attribute since Python 3. `args[0]` is the conventional error message but even that is not entirely true: for OSError, `args[0]` is the errno, `args[1]` is the strerror. `str(error)`
